packages/back-end/src/services/datasource.ts

```diff
@@ -101,6 +108,38 @@ export function getSourceIntegrationObject(datasource: DataSourceInterface) {
   return obj;
 }
 
+export async function generateInformationSchema(
+  datasource: DataSourceInterface
+): Promise<
+  | {
+      informationSchema: InformationSchema[];
+      error?: undefined;
+    }
+  | { error: string; informationSchema?: undefined }
```

Comment:
Is it required to specify `error?: undefined` in the success case and `informationSchema?: undefined` in the error case? I didn't think it was but maybe I'm mistaken.

Reply:
It doesn't look like we need the error since we aren't using it. Could we simplify this method with something more inline with how we handle other methods (like model methods) where we return either the record or null, and then let the caller handle the empty case? Here's an example:

```ts
export async function generateInformationSchema(
  datasource: DataSourceInterface
): Promise<InformationSchema[] | null> {
  const integration = getSourceIntegrationObject(datasource);

  try {
    return await integration.getInformationSchema(
      datasource.type,
      integration.params.projectId
    );
  } catch (e) {
    return null
  }
}
```

Main reasons: 

- `getSourceIntegrationObject()` seems to always return an object, otherwise it'll throw. We won't even get to the `if` if there's no source integration object so it seems redundant.
- if all integrations have `getInformationSchema()` and return either a list of information schema or null, then we don't need to check for `getInformationSchema` being present. This is related to the feedback I added before where we can add this to the abstract class and have it return null by default.

Some reasons we may not want to implement this feedback:

- we want to know the error, in which case is returning it the best option, or should we log it to Sentry?

Reply:
@tinahollygb Thank you for clarifying, I just updated the if statement above to remove the unnecessary `!integration` check.

Re: Adding the `getInformationSchema` to the base class and have it return null by default sounds interesting. Currently, and as far as I know, each datasource type that we'll ultimately add support for `informationSchema` will require a slightly different method of obtaining the informationSchema.

Were you thinking another switch case statement inside of that `getInformationSchema` method in order to support that or a different pattern?

Reply:
No, I wasn't thinking of a switch in `getInformationSchema` itself but perhaps whoever calls it if that's necessary.

I had previously provided example code for this: https://github.com/growthbook/growthbook/pull/940#discussion_r1120762848

Let me know if you had any questions or something about that isn't clear.

Reply:
Further example of this is also in this comment too: https://github.com/growthbook/growthbook/pull/940#discussion_r1092555417

Reply:
@tinahollygb I just pushed up a commit [here](https://github.com/growthbook/growthbook/pull/940/commits/a8e3b8d6a3bc08ee9edf95a30c839a238189a06a) that moves the `getInformationSchema` to the `SqlIntegration` base class.

I don't particularly love it, and I'm not 100% convinced this is actually what you were suggesting.

In one of the examples in [comment #940](https://github.com/growthbook/growthbook/pull/940#discussion_r1092555417) had an example of calling `handleBigQueryInformationSchema(informationSchema)`, but at that point, `informationSchema` hasn't yet been defined - the `handleBigQueryInformationSchema` method's purpose is to build the `informationSchema`.

Ultimately, I moved `getBigQueryInformationSchema` and `getPostgresInformationSchema` method into the `SqlIntegration` class, rather than in their respective classes to make Typescript happy. But if this is the direction you were suggesting, I can explore that further, I just wanted to get this commit up to get your feedback before I spent more time on that.

I really appreciate the time you've invested into this today, and if it'd be more efficient to talk through this over a quick chat, I'm happy to do that, though, I can't until tomorrow.